src/librustc_mir/borrow_check/nll/invalidation.rs

```diff
@@ -0,0 +1,955 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use borrow_check::borrow_set::{BorrowSet, BorrowData};
+use borrow_check::location::LocationTable;
+use borrow_check::{JustWrite, WriteAndRead};
+use borrow_check::{ShallowOrDeep, Deep, Shallow};
+use borrow_check::{ReadOrWrite, Activation, Read, Reservation, Write};
+use borrow_check::{Context, ContextKind};
+use borrow_check::{LocalMutationIsAllowed, MutateMode};
+use borrow_check::ArtificialField;
+use borrow_check::{ReadKind, WriteKind, Overlap};
+use borrow_check::nll::facts::AllFacts;
+use dataflow::move_paths::indexes::BorrowIndex;
+use rustc::hir;
+use rustc::hir::def_id::DefId;
+use rustc::infer::InferCtxt;
+use rustc::mir::visit::Visitor;
+use rustc::mir::{BasicBlock, Location, Mir, Place, Rvalue, Projection};
+use rustc::mir::{Local, ProjectionElem};
+use rustc::mir::{Statement, StatementKind};
+use rustc::mir::{Terminator, TerminatorKind};
+use rustc::mir::{Field, Operand, BorrowKind};
+use rustc::ty::{self, ParamEnv};
+use rustc_data_structures::indexed_vec::Idx;
+use std::iter;
+
+pub(super) fn generate_invalidates<'cx, 'gcx, 'tcx>(
+    infcx: &InferCtxt<'cx, 'gcx, 'tcx>,
+    all_facts: &mut Option<AllFacts>,
+    location_table: &LocationTable,
+    mir: &Mir<'tcx>,
+    mir_def_id: DefId,
+    borrow_set: &BorrowSet<'tcx>,
+) {
+    if !all_facts.is_some() {
```

Comment:
Nit: I'd probably write:

```rust
if let Some(all_facts) = all_facts {
    .. // inner stuff
}
```

then you don't, I think, need the call to `take()` etc. But whatever.